pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,353 @@
+package ftp
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/textproto"
+	"os"
+	"path"
+	"time"
+
+	"github.com/jlaffaye/ftp"
+)
+
+// Conn struct embeds the *ftp.ServerConn returned by ftp server on successful connection.
+type Conn struct {
+	*ftp.ServerConn
+}
+
+// Retr wraps the ftp retrieve method to return a ftpResponse interface type.
+func (c *Conn) Retr(filepath string) (ftpResponse, error) {
+	return c.ServerConn.Retr(filepath)
+}
+
+func (c *Conn) RetrFrom(filepath string, offset uint64) (ftpResponse, error) {
+	return c.ServerConn.RetrFrom(filepath, offset)
+}
+
+var (
+	errEmptyFilename            = errors.New("filename cannot be empty")
+	errEmptyPath                = errors.New("file/directory path cannot be empty")
+	errEmptyDirectory           = errors.New("directory name cannot be empty")
+	errInvalidArg               = errors.New("invalid filename/directory")
+	transferCompleteError       = &textproto.Error{Code: 226, Msg: "Transfer complete."}
+	directoryAlreadyExistsError = &textproto.Error{Code: 550, Msg: "Create directory operation failed."}
+)
+
+// ftpFileSystem represents a file system interface over FTP.
+type ftpFileSystem struct {
+	*ftpFile
+	conn    ServerConn
+	config  *Config
+	logger  Logger
+	metrics Metrics
+}
+
+// Config represents the FTP configuration.
+type Config struct {
+	Host      string // FTP server hostname
+	User      string // FTP username
+	Password  string // FTP password
+	Port      int    // FTP port
+	RemoteDir string // Remote directory path. Base Path for all FTP Operations.
+}
+
+// New initializes a new instance of ftpFileSystem with provided configuration.
+func New(config *Config) FileSystemProvider {
+	return &ftpFileSystem{config: config}
+}
+
+// UseLogger sets the logger interface for the FTP file system.
+func (f *ftpFileSystem) UseLogger(logger interface{}) {
+	if l, ok := logger.(Logger); ok {
+		f.logger = l
+	}
+}
+
+// UseMetrics sets the metrics for the MongoDB client which asserts the Metrics interface.
+func (f *ftpFileSystem) UseMetrics(metrics interface{}) {
+	if m, ok := metrics.(Metrics); ok {
+		f.metrics = m
+	}
+}
+
+// Connect establishes a connection to the FTP server and logs in.
+func (f *ftpFileSystem) Connect() {
+	ftpServer := fmt.Sprintf("%v:%v", f.config.Host, f.config.Port)
+
+	defer f.processLog(&FileLog{Operation: "Connect", Location: ftpServer}, time.Now())
+
+	const dialTimeout = 5 * time.Second
```

Comment:
This should be a configuration variable

Reply:
We will be not providing it as a part of configs for now.